File: csharp/dms-service-sdk-test/Test/PaymentClient.cs
using Dms.Service.Sdk.Event;
using Nethereum.Util;

namespace Dms.Service.Sdk.Test;

using Client;
using Types;
using Utils;


public class PaymentClientTest
{
    private PaymentClient paymentClient;
    private PaymentClientForUser userClient;
    private PaymentClientForShop shopClient;
    private TestEventListener listener;
    private TaskEventCollector collector;
    private string temporaryAccount;
    private string terminalID = "POS001";
    private PaymentTaskItem paymentItem;

    public PaymentClientTest()
    {
        string privateKeyForPayment = "0x8acceea5937a8e4bb07abc93a1374264dd9bd2fc384c979717936efe63367276";
        paymentClient = new PaymentClient(NetWorkType.TestNet, privateKeyForPayment);
        listener = new TestEventListener();
        collector = new TaskEventCollector(paymentClient, listener);
        collector.Start();
        userClient = new PaymentClientForUser(NetWorkType.TestNet, 
            "0x70438bc3ed02b5e4b76d496625cb7c06d6b7bf4362295b16fdfe91a046d4586c");
        shopClient = new PaymentClientForShop(NetWorkType.TestNet,
            "0xa237d68cbb66fd5f76e7b321156c46882546ad87d662dec8b82703ac31efbf0a",
            "0x0001be96d74202df38fd21462ffcef10dfe0fcbd7caa3947689a3903e8b6b874");
    }
    
    [SetUp]
    public void Setup()
    {
    }

    [Test]
    public void Test01_Waiting()
    {
        Thread.Sleep(3000);
    }

    [Test]
    public void Test02_CreateTemporaryAccount()
    {
        temporaryAccount = userClient.GetTemporaryAccount();
        System.Console.WriteLine($"Temporary Account: {temporaryAccount}");
    }

    [Test]
    public void Test03_OpenNewPayment()
    {
        var purchaseId = CommonUtils.GetSamplePurchaseId();
        paymentItem = paymentClient.OpenNewPayment(
            purchaseId,
            temporaryAccount,
            Amount.Make("1_000").Value,
            "php",
            shopClient.ShopId,
            terminalID
        );
        Assert.That(paymentItem.PurchaseId, Is.EqualTo(purchaseId));
        Assert.That(paymentItem.Account, Is.EqualTo(userClient.Address));
    }
    
    [Test]
    public void Test04_Waiting()
    {
        Thread.Sleep(1000);
    }
    
    [Test]
    public void Test05_ApproveNewPayment()
    {
        var res = userClient.ApproveNewPayment(
            paymentItem.PaymentId,
            paymentItem.PurchaseId,
            paymentItem.Amount,
            paymentItem.Currency,
            paymentItem.ShopId,
            true
        );
        Assert.That(res.PaymentId, Is.EqualTo(paymentItem.PaymentId));
    }

    [Test]
    public void Test06_Waiting()
    {
        Thread.Sleep(3000);
    }
    
    [Test]
    public void Test07_CloseNewPayment()
    {
        var res = paymentClient.CloseNewPayment(
            paymentItem.PaymentId,
            true
        );
        Assert.That(res.PaymentId, Is.EqualTo(paymentItem.PaymentId));
    }

    [Test]
    public void Test08_Waiting()
    {
        Thread.Sleep(3000);
    }

    [Test]
    public void Test09_OpenCancelPayment()
    {
        var res = paymentClient.OpenCancelPayment(
            paymentItem.PaymentId,
            terminalID
        );
        Assert.That(res.PurchaseId, Is.EqualTo(paymentItem.PurchaseId));
        Assert.That(res.Account, Is.EqualTo(userClient.Address));
    }
    
    [Test]
    public void Test10_Waiting()
    {
        Thread.Sleep(1000);
    }
    
    [Test]
    public void Test11_ApproveCancelPayment()
    {
        var res = shopClient.ApproveCancelPayment(
            paymentItem.PaymentId,
            paymentItem.PurchaseId,
            true
        );
        Assert.That(res.PaymentId, Is.EqualTo(paymentItem.PaymentId));
    }

    [Test]
    public void Test12_Waiting()
    {
        Thread.Sleep(3000);
    }
    
    [Test]
    public void Test13_CloseCancelPayment()
    {
        var res = paymentClient.CloseCancelPayment(
            paymentItem.PaymentId,
            true
        );
        Assert.That(res.PaymentId, Is.EqualTo(paymentItem.PaymentId));
    }

    [Test]
    public void Test14_Waiting()
    {
        Thread.Sleep(3000);
        collector.Stop();
    }
}

File: java/sdk/src/test/java/org/acc/service/sdk/client/PaymentClient.java
package org.acc.service.sdk.client;

import org.acc.service.sdk.data.NetWorkType;
import org.acc.service.sdk.data.payment.PaymentInfo;
import org.acc.service.sdk.data.payment.PaymentTaskItem;
import org.acc.service.sdk.data.payment.ShopTaskItem;
import org.acc.service.sdk.event.ITaskEventListener;
import org.acc.service.sdk.event.TaskEventCollector;
import org.acc.service.sdk.utils.Amount;
import org.acc.service.sdk.utils.CommonUtils;
import org.junit.jupiter.api.Test;

import java.math.BigInteger;

import static org.junit.jupiter.api.Assertions.assertEquals;

class PaymentClientTest {
    @Test
    void PaymentClient() {
        try {
            // Create Client
            System.out.println("[ Create Client ]");
            String privateKeyForPayment = "0x8acceea5937a8e4bb07abc93a1374264dd9bd2fc384c979717936efe63367276";
            PaymentClient client = new PaymentClient(NetWorkType.testnet, privateKeyForPayment);

            // Create Event Collector
            System.out.println("[ Create Event Collector ]");
            TaskEventListener listener = new TaskEventListener();
            TaskEventCollector collector = new TaskEventCollector(client, listener);

            // Create User Client
            System.out.println("[ Create User Client ]");
            PaymentClientForUser userClient = new PaymentClientForUser(NetWorkType.testnet, "0x70438bc3ed02b5e4b76d496625cb7c06d6b7bf4362295b16fdfe91a046d4586c");

            // Create Shop Client
            System.out.println("[ Create Shop Client ]");
            PaymentClientForShop shopClient = new PaymentClientForShop(NetWorkType.testnet, "0xa237d68cbb66fd5f76e7b321156c46882546ad87d662dec8b82703ac31efbf0a", "0x0001be96d74202df38fd21462ffcef10dfe0fcbd7caa3947689a3903e8b6b874");

            // Start Event Collector
            collector.start();
            Thread.sleep(3000);

            // Create Temporary Account
            System.out.println("[ Create Temporary Account ]");
            String temporaryAccount = userClient.getTemporaryAccount();
            System.out.printf("  - Temporary Account: %s\n", temporaryAccount);

            String terminalID = "POS001";

            //  Get Payment Info
            PaymentInfo value = client.getPaymentInfo("0x64D111eA9763c93a003cef491941A011B8df5a49", BigInteger.valueOf(100000000), "php");
            assertEquals(value.account, "0x64D111eA9763c93a003cef491941A011B8df5a49");

            // Open New Payment
            System.out.println("[ Open New Payment ]");
            PaymentTaskItem paymentItem = client.openNewPayment(
                    CommonUtils.getSamplePurchaseId(),
                    temporaryAccount,
                    Amount.make("1_000").getValue(),
                    "php",
                    shopClient.getShopId(),
                    terminalID
            );

            // Waiting...
            System.out.println("[ Waiting... ]");
            Thread.sleep(3000);

            // Approval New Payment
            System.out.println("[ Approval New Payment ]");
            userClient.approveNewPayment(
                    paymentItem.paymentId,
                    paymentItem.purchaseId,
                    paymentItem.amount,
                    paymentItem.currency,
                    paymentItem.shopId,
                    true
            );

            // Waiting...
            System.out.println("[ Waiting... ]");
            Thread.sleep(3000);

            // Close New Payment
            System.out.println("[ Close New Payment ]");
            client.closeNewPayment(paymentItem.paymentId, true);

            // Waiting...
            System.out.println("[ Waiting... ]");
            Thread.sleep(3000);

            // Open Cancel Payment
            System.out.println("[ Open Cancel Payment ]");
            client.openCancelPayment(paymentItem.paymentId, terminalID);

            // Waiting...
            System.out.println("[ Waiting... ]");
            Thread.sleep(3000);

            // Approval Cancel Payment
            System.out.println("[ Approval Cancel Payment ]");
            shopClient.approveCancelPayment(
                    paymentItem.paymentId,
                    paymentItem.purchaseId,
                    true
            );

            // Waiting...
            System.out.println("[ Waiting... ]t");
            Thread.sleep(3000);

            // Close Cancel Payment
            System.out.println("[ Close Cancel Payment ]");
            client.closeCancelPayment(paymentItem.paymentId, true);

            PaymentTaskItem paymentTaskItem = client.getPaymentItem(paymentItem.paymentId);
            assertEquals(paymentTaskItem.paymentId, paymentItem.paymentId);

            // Waiting...
            System.out.println("[ Waiting... ]");
            Thread.sleep(3000);

            // Stop Event Collector
            Thread.sleep(3000);
            collector.stop();

        } catch (Exception e) {
            assertEquals("some exception message...", e.getMessage());
        }
    }
}

class TaskEventListener implements ITaskEventListener {
    public void onNewPaymentEvent(
            String type,
            int code,
            String message,
            long sequence,
            PaymentTaskItem paymentTaskItem
    ) {
        System.out.printf("  -> onNewPaymentEvent %s - %d - %s - %d\n", type, code, message, sequence);
    }

    public void onNewShopEvent(
            String type,
            int code,
            String message,
            long sequence,
            ShopTaskItem shopTaskItem
    ) {
        System.out.printf("  -> onNewShopEvent %s - %d - %s - %d\n", type, code, message, sequence);
    }
}
